### ISSUE TITLE
Upgrade from 9.3.1 to 9.3.11, 9.4.9, or 9.5.1 fails on updating alert configuration and/or secret migration

### ISSUE THREAD

**Issue by andrewbierbaum** [grafana/grafana#67316]
## What happened:
In trying to bump the grafana version from 9.3.1 to 9.3.11, 9.4.9, or 9.5.1, secret migration issues appear

## Environment:
- Grafana version: 9.3.1
- OS Grafana is installed on: focal docker container

When attempting to upgrade to 9.5.1, I got the following log `logger=ngalert t=2023-04-26T17:31:01.693020413Z level=warn msg="Unexpected number of rows updating alert configuration history" rows=0 org=1 hash=not-yet-calculated`

Attempting to upgrade to 9.4.9 or 9.3.11, I'm seeing `migrations completed" performed=0 skipped=464 duration=760.3µs` and grafana appears to crash immediately after `starting MultiOrg Alertmanager`, and attempt to come back up but fails with `Server lock for secret migration already exists`, 
```
logger=settings t=2023-04-20T17:33:31.648725608Z level=info msg="Starting Grafana" version=9.3.11 commit=f9548d33f8 branch=HEAD compiled=2023-03-16T23:16:52Z
logger=settings t=2023-04-20T17:33:31.648969125Z level=info msg="Config loaded from" file=/usr/share/grafana/conf/defaults.ini
logger=settings t=2023-04-20T17:33:31.648989404Z level=info msg="Config loaded from" file=/usr/share/grafana/conf/custom.ini
logger=settings t=2023-04-20T17:33:31.649002016Z level=info msg="Config overridden from command line" arg="default.paths.data=/var/lib/grafana"
logger=settings t=2023-04-20T17:33:31.649028961Z level=info msg="Config overridden from command line" arg="default.paths.logs=/var/log/grafana"
logger=settings t=2023-04-20T17:33:31.649042806Z level=info msg="Config overridden from command line" arg="default.paths.plugins=/var/lib/grafana/plugins"
logger=settings t=2023-04-20T17:33:31.649057692Z level=info msg="Config overridden from Environment variable" var="GF_SERVER_ROOT_URL=http://localhost:3000"
logger=settings t=2023-04-20T17:33:31.649070067Z level=info msg="Config overridden from Environment variable" var="GF_DATABASE_TYPE=mysql"
logger=settings t=2023-04-20T17:33:31.649084517Z level=info msg="Config overridden from Environment variable" var="GF_DATABASE_HOST=mysqlhost1:3306"
logger=settings t=2023-04-20T17:33:31.649106276Z level=info msg="Config overridden from Environment variable" var="GF_DATABASE_USER=jenga"
logger=settings t=2023-04-20T17:33:31.649131802Z level=info msg="Config overridden from Environment variable" var="GF_DATABASE_PASSWORD=*********"
logger=settings t=2023-04-20T17:33:31.649145485Z level=info msg="Path Home" path=/usr/share/grafana
logger=settings t=2023-04-20T17:33:31.649157602Z level=info msg="Path Data" path=/var/lib/grafana
logger=settings t=2023-04-20T17:33:31.649177191Z level=info msg="Path Logs" path=/var/log/grafana
logger=settings t=2023-04-20T17:33:31.649200155Z level=info msg="Path Plugins" path=/grafana/plugins
logger=settings t=2023-04-20T17:33:31.649224435Z level=info msg="Path Provisioning" path=/grafana/dockers/grafana/provisioning/
logger=settings t=2023-04-20T17:33:31.649286965Z level=info msg="App mode production"
logger=sqlstore t=2023-04-20T17:33:31.649408736Z level=info msg="Connecting to DB" dbtype=mysql
logger=migrator t=2023-04-20T17:33:31.675061307Z level=info msg="Starting DB migrations"
logger=migrator t=2023-04-20T17:33:31.683121179Z level=info msg="migrations completed" performed=0 skipped=464 duration=760.3µs
logger=plugin.loader t=2023-04-20T17:33:31.85492733Z level=info msg="Plugin registered" pluginID=input
logger=plugin.loader t=2023-04-20T17:33:33.852229977Z level=info msg="Plugin registered" pluginID=grafana-piechart-panel
logger=plugin.loader t=2023-04-20T17:33:33.852339107Z level=info msg="Plugin registered" pluginID=grafana-simple-json-datasource
logger=plugin.loader t=2023-04-20T17:33:33.852373633Z level=info msg="Plugin registered" pluginID=alexanderzobnin-zabbix-datasource
logger=plugin.alexanderzobnin-zabbix-datasource t=2023-04-20T17:33:33.860334118Z level=info msg=Profiler enabled=false
logger=plugin.alexanderzobnin-zabbix-datasource t=2023-04-20T17:33:33.86205971Z level=warn msg="Plugin process is running with elevated privileges. This is not recommended"
logger=plugin.loader t=2023-04-20T17:33:33.862096542Z level=info msg="Plugin registered" pluginID=alexanderzobnin-zabbix-triggers-panel
logger=plugin.loader t=2023-04-20T17:33:33.86211229Z level=info msg="Plugin registered" pluginID=marcusolsson-json-datasource
logger=plugin.loader t=2023-04-20T17:33:33.862125056Z level=info msg="Plugin registered" pluginID=alexanderzobnin-zabbix-app
logger=plugin.loader t=2023-04-20T17:33:33.862137241Z level=info msg="Plugin registered" pluginID=blackmirror1-singlestat-math-panel
logger=plugin.loader t=2023-04-20T17:33:33.862151549Z level=info msg="Plugin registered" pluginID=boazreicher-sierraplot-panel
logger=plugin.loader t=2023-04-20T17:33:33.862164116Z level=info msg="Plugin registered" pluginID=flant-statusmap-panel
logger=secrets t=2023-04-20T17:33:33.862589671Z level=info msg="Envelope encryption state" enabled=true currentprovider=secretKey.v1
logger=query_data t=2023-04-20T17:33:33.865077331Z level=info msg="Query Service initialization"
logger=live.push_http t=2023-04-20T17:33:33.868190477Z level=info msg="Live Push Gateway initialization"
logger=infra.usagestats.collector t=2023-04-20T17:33:34.474727565Z level=info msg="registering usage stat providers" usageStatsProvidersLen=2
logger=server t=2023-04-20T17:33:34.474918457Z level=info msg="Writing PID file" path=/highland/grafana/tmp/pids/server.pid pid=124
logger=provisioning.alerting t=2023-04-20T17:33:43.690347737Z level=warn msg="file has invalid suffix 'README' (.yaml,.yml,.json accepted), skipping"
logger=provisioning.alerting t=2023-04-20T17:33:43.691062039Z level=info msg="starting to provision alerting"
logger=provisioning.alerting t=2023-04-20T17:33:43.691090635Z level=info msg="finished to provision alerting"
logger=ngalert.state.manager t=2023-04-20T17:33:43.691494596Z level=info msg="Warming state cache for startup"
logger=grafanaStorageLogger t=2023-04-20T17:33:43.691836497Z level=info msg="storage starting"
logger=secret.migration t=2023-04-20T17:33:43.693569288Z level=error msg="Server lock for secret migration already exists"
logger=http.server t=2023-04-20T17:33:43.693766389Z level=info msg="HTTP Server Listen" address=[::]:3000 protocol=http subUrl= socket=
logger=ngalert.state.manager t=2023-04-20T17:33:43.696980899Z level=info msg="State cache has been initialized" states=0 duration=5.481082ms
logger=ticker t=2023-04-20T17:33:43.69705647Z level=info msg=starting first_tick=2023-04-20T17:33:50Z
logger=ngalert.multiorg.alertmanager t=2023-04-20T17:33:43.69709042Z level=info msg="starting MultiOrg Alertmanager"
mysql: [Warning] Using a password on the command line interface can be insecure.
mysql: [Warning] Using a password on the command line interface can be insecure.
Grafana server is running with elevated privileges. This is not recommended
```
and then I see grafana attempt to start again, but fail because the above with Server lock for secret migration already exists and the pervious 0.0.0.0:3000 is already in use.
```
logger=settings t=2023-04-20T17:33:45.073958312Z level=info msg="Starting Grafana" version=9.3.11 commit=f9548d33f8 branch=HEAD compiled=2023-03-16T23:16:52Z

logger=secret.migration t=2023-04-20T17:33:57.691036258Z level=error msg="Server lock for secret migration already exists"
logger=server t=2023-04-20T17:33:57.692481377Z level=error msg="Stopped background service" service=*api.HTTPServer reason="failed to open listener on address 0.0.0.0:3000: listen tcp 0.0.0.0:3000: bind: address already in use"
I'm happy to grab more information or to start another issue tread if this is considered too different.
```

**Comment by tonypowa** [comment]
back-end squad: please reroute if needed

**Comment by tolzhabayev** [comment]
@grafana/grafana-authnz-team could you please take a look at the lock around secret migration?

**Comment by tolzhabayev** [comment]
@grafana/grafana-operator-experience-squad please take a look

**Comment by leandro-deveikis** [comment]
hi!
I will send this to the alerting team internally as I can not find the GH team for it. 
It seems that first, the alerting configuration migration failed somehow, and that left the database in an inconsistent state. After that, the server was unable to start up. 

About the secrets migration lock, it has a timeout of 10 minutes. From the logs, I see that time has not passed between the attempts to start up. So it should resolve itself waiting that time. If that problem persists, let me know and I can provide steps to delete the lock (just a row in the database) but would recommend waiting for the alerting team to respond so we don't keep doing changes to the database.

**Comment by andrewbierbaum** [comment]
@leandro-deveikis I attempted stopping the service for longer than 10 minutes, and it still did not start up properly. Please provide steps or the location of the secrets migration lock. This grafana is backed by a mysql host.

**Comment by leandro-deveikis** [comment]
@andrewbierbaum Can you provide the logs for the last attempt to run? Even if the migration is still locked, it shouldn't prevent from the grafana starting up or working (if those are not migrated, should use the old ones). 

Still, if you want to check, in the database there is a table called server_lock, and it should be a row there with the operation_uid=SecretMigrationService. Deleting that row would release the lock

**Comment by dafydd-t** [comment]
I've added the alerting label to get more relevant eyes on this 👍

**Comment by santihernandezc** [comment]
Hi @andrewbierbaum, those _"Unexpected number of rows updating alert configuration history"_ logs you're referring to shouldn't be a concern. It's just a warning message describing expected behavior, not an error.

Grafana does not support migrating to older versions. Once migrations for a given version run, the database is not guaranteed to be in a state that is readable for older versions of the system.

My recommendation would be to just run Grafana v9.5.1. Please write back to us should you run into further issues.

**Comment by nice-pink** [comment]
I'm having the same issue. Problem is, that the upgrade fails with the above logs. So the recommendation to the run the newest version is what I'm trying to achieve, but fails.

**Comment by andrewbierbaum** [comment]
Sorry for the slow reply. Agreed, I would like to run 9.3.11, 9.4.9, or 9.5.1, but any upgrade fails, and I have to go back to 9.3.1

**Comment by andrewbierbaum** [comment]
I tried again to upgrade from 9.3.1 to 9.5.1 with debug logging and I did find a log that may be helpful

```
logger=server t=2023-05-03T15:49:56.056770446Z level=debug msg="NOTIFY_SOCKET environment variable empty or unset, can't send systemd notification"
logger=serviceaccounts serviceaccounts.background=(MISSING) t=2023-05-03T15:49:56.056794561Z level=debug msg="service initialized"
logger=server t=2023-05-03T15:49:56.05680595Z level=debug msg="Waiting on services..."
logger=server t=2023-05-03T15:49:56.056762343Z level=debug msg="Stopped background service" service=*store.dummyEntityEventsService reason=null
logger=server t=2023-05-03T15:49:56.056784946Z level=debug msg="Starting background service" service=*loginattemptimpl.Service
logger=server t=2023-05-03T15:49:56.056767384Z level=debug msg="Starting background service" service=*authinfoservice.Implementation
logger=server t=2023-05-03T15:49:56.056768633Z level=debug msg="Starting background service" service=*process.Manager
logger=infra.lockservice t=2023-05-03T15:49:56.056849853Z level=debug msg="Start LockExecuteAndRelease" actionName="secret migration task "
logger=server t=2023-05-03T15:49:56.056785524Z level=debug msg="Starting background service" service=*supportbundlesimpl.Service
logger=login.authinfo t=2023-05-03T15:49:56.056913552Z level=debug msg="Started AuthInfo Metrics collection service"
logger=server t=2023-05-03T15:49:56.057488631Z level=error msg="Stopped background service" service=*api.HTTPServer reason="failed to open listener on address 0.0.0.0:3000: listen tcp 0.0.0.0:3000: bind: address already in use"
```

I'm trying to continue digging in the logs and output them now

**Comment by mmandrus** [comment]
Hello,

The last line: `logger=server t=2023-05-03T15:49:56.057488631Z level=error msg="Stopped background service" service=*api.HTTPServer reason="failed to open listener on address 0.0.0.0:3000: listen tcp 0.0.0.0:3000: bind: address already in use"`

Suggests that you already have a Grafana process running in the background, maybe one that didn't terminate properly. Have you tried looking for the stale process and killing it, or rebooting? If you have another process using this port, you can also try changing the Grafana port by adding this to your config.ini file:

```
[server]
http_port = 1234 ; your port here
```

**Comment by andrewbierbaum** [comment]
I tested again trying to delete and create the grafana database, and it is still failing to start in a similar fashion
```
mysql> drop database grafana;
Query OK, 63 rows affected (0.52 sec)

mysql> create database grafana;
Query OK, 1 row affected (0.01 sec)
```

**Comment by papagian** [comment]
@andrewbierbaum I can still see the same error (with the new port):
```
logger=server t=2023-05-03T17:33:05.087967938Z level=error msg="Stopped background service" service=*api.HTTPServer reason="failed to open listener on address 0.0.0.0:3001: listen tcp 0.0.0.0:3001: bind: address already in use"
```

how do you start the container? do you use any custom configuration?
also, can you check which process is listening to `3001` port?

**Comment by andrewbierbaum** [comment]
it appears the way systemctl starts grafana changed between 9.3.1 and the newest versions. I quickly attempted a change, and ran into some accumulated tech debt we had built up over the years. 

I ended up just using the upstream grafana docker image to make future maintainability more straightforward. Thanks for the help.
